### PR TITLE
cdist performance improvement for euclidean distance

### PR DESCRIPTION
**Pull request by ifedan**
@jacobrgardner https://github.com/pytorch/pytorch/issues/15253#issuecomment-491467128 preposed a way to speedup euclidean distance calculation. This PR is implementation of this solution for normal and batch version.

Also @simonepri provided performance metrics https://github.com/pytorch/pytorch/issues/15253#issuecomment-502363581
image

Current implementation has speedup comparing to @jacobrgardner approach
image

### CODE REVIEW

**Comment by jacobrgardner** [comment]
Awesome!

Just for completeness, here is a self contained version of our current distance function in gpytorch, which is about 2x faster than the original snippet I posted in that issue:
```python
def fast_cdist(x1, x2):
    adjustment = x1.mean(-2, keepdim=True)
    x1 = x1 - adjustment
    x2 = x2 - adjustment  # x1 and x2 should be identical in all dims except -2 at this point

    # Compute squared distance matrix using quadratic expansion
    # But be clever and do it with a single matmul call
    x1_norm = x1.pow(2).sum(dim=-1, keepdim=True)
    x1_pad = torch.ones_like(x1_norm)
    x2_norm = x2.pow(2).sum(dim=-1, keepdim=True)
    x2_pad = torch.ones_like(x2_norm)
    x1_ = torch.cat([-2. * x1, x1_norm, x1_pad], dim=-1)
    x2_ = torch.cat([x2, x2_pad, x2_norm], dim=-1)
    res = x1_.matmul(x2_.transpose(-2, -1))

    # Zero out negative values
    res.clamp_min_(1e-30).sqrt_()
    return res
```
This uses the same basic strategy as the original method I'd written, but does everything in a single `matmul` call because matmul calls are really just that fast.

That being said, it has some drawbacks for inclusion upstream in pytorch proper over my original snippet. Namely (1) the additional working memory overhead, and (2) it's the same basic strategy, so a proper implementation of the first snippet I'd linked may erase the gap anyways.

**Comment by ifedan** [comment]
> matmul



> Awesome!
> 
> Just for completeness, here is a self contained version of our current distance function in gpytorch, which is about 2x faster than the original snippet I posted in that issue:
> 
> ```python
> def fast_cdist(x1, x2):
>     adjustment = x1.mean(-2, keepdim=True)
>     x1 = x1 - adjustment
>     x2 = x2 - adjustment  # x1 and x2 should be identical in all dims except -2 at this point
> 
>     # Compute squared distance matrix using quadratic expansion
>     # But be clever and do it with a single matmul call
>     x1_norm = x1.pow(2).sum(dim=-1, keepdim=True)
>     x1_pad = torch.ones_like(x1_norm)
>     x2_norm = x2.pow(2).sum(dim=-1, keepdim=True)
>     x2_pad = torch.ones_like(x2_norm)
>     x1_ = torch.cat([-2. * x1, x1_norm, x1_pad], dim=-1)
>     x2_ = torch.cat([x2, x2_pad, x2_norm], dim=-1)
>     res = x1_.matmul(x2_.transpose(-2, -1))
> 
>     # Zero out negative values
>     res.clamp_min_(1e-30).sqrt_()
>     return res
> ```
> 
> This uses the same basic strategy as the original method I'd written, but does everything in a single `matmul` call because matmul calls are really just that fast.
> 
> That being said, it has some drawbacks for inclusion upstream in pytorch proper over my original snippet. Namely (1) the additional working memory overhead, and (2) it's the same basic strategy, so a proper implementation of the first snippet I'd linked may erase the gap anyways.

@jacobrgardner Thanks, let me check the performance metrics.

**Comment by simonepri** [comment]
Great work!

FYI, I just added a reference to an implementation for the batched version in https://github.com/pytorch/pytorch/issues/15253#issuecomment-529022752

**Comment by simonepri** [comment]
Great work!

FYI, I just added a reference to an implementation for the batched version in https://github.com/pytorch/pytorch/issues/15253#issuecomment-529022752

cc: @lerks

File: aten/src/ATen/native/Distance.cpp [hunk 1]
```diff
@@ -24,6 +24,20 @@ Tensor pdist(const Tensor& self, const double p) {
...
+Tensor euclidean_dist_out(const Tensor& x1, const Tensor& x2) {
+  Tensor x1_norm = x1.pow(2).sum(-1, true);
+  Tensor x2_norm = x2.pow(2).sum(-1, true);
+  Tensor result;
+  if (x1.dim() == 2) {
+    result = at::addmm(x2_norm.transpose(-2, -1), x1, x2.transpose(-2, -1), 1, -2);
+  } else {
+    result = at::baddbmm(x2_norm.transpose(-2, -1), x1, x2.transpose(-2, -1), 1, -2);
+  }
+  result.add_(x1_norm);
```
Comment (ngimel) [inline aten/src/ATen/native/Distance.cpp]: Implementations in #15253 had a call filtering negative values, and it is necessary, it is possible to get negative distances with this approach because of floating point accuracy.

Comment (ifedan) [reply aten/src/ATen/native/Distance.cpp]: yep, I changed this

**Comment by Balandat** [comment]
@ifedan any thoughts on being able to return the squared (or more generally `p`-ed) distance?

That is have an option to return `\sum_k (x_ik - xj_k)^p` instead of   `(\sum_k (x_ik - xj_k)^p(^(1/p)`.

 This is not just as easy as squaring the result of cdist, as cdist is not differentiable at zero.

**Comment by ifedan** [comment]
**Speedup**
image

**Comment by ifedan** [comment]
> @ifedan any thoughts on being able to return the squared (or more generally `p`-ed) distance?
> 
> That is have an option to return `\sum_k (x_ik - xj_k)^p` instead of `(\sum_k (x_ik - xj_k)^p(^(1/p)`.
> 
> This is not just as easy as squaring the result of cdist, as cdist is not differentiable at zero.

I think it can be a separate function, something like euclidian_distance, because cdist also compute other p-distances

**Review by ngimel — APPROVED** [review]

**Review by houseroad — APPROVED** [review]
Adding an optional parameter with default value is backward compatible. Looks good to me.

**Comment by promach** [comment]
What do you guys think about https://stackoverflow.com/questions/61241523/replacing-torch-cdist-function-to-eliminate-gpu-out-of-memory-runtime-error ?

```
def new_cdist(p, eta):
    class cdist(torch.autograd.Function):
        @staticmethod
        def forward(ctx, W, X):
            ctx.save_for_backward(W, X)
            out = -torch.cdist(W, X, p)
            return out

        @staticmethod
        def backward(ctx, grad_output):
            W, X = ctx.saved_tensors
            grad_W = grad_X = None
            if ctx.needs_input_grad[0]:
                _temp1 = torch.unsqueeze(X, 2).expand(X.shape[0], X.shape[1], W.shape[0]).permute(1, 0, 2)
                _temp2 = torch.unsqueeze(W.transpose(0, 1), 1)
                _temp = torch.cdist(_temp1, _temp2, p).squeeze().transpose(0, 1)
                grad_W = torch.matmul(grad_output, _temp)
                # print('before norm: ', torch.norm(grad_W))
                grad_W = eta * np.sqrt(grad_W.numel()) / torch.norm(grad_W) * grad_W
                print('after norm: ', torch.norm(grad_W))
            if ctx.needs_input_grad[1]:
                _temp1 = torch.unsqueeze(W, 2).expand(W.shape[0], W.shape[1], X.shape[0]).permute(1, 0, 2)
                _temp2 = torch.unsqueeze(X.transpose(0, 1), 1)
                _temp = torch.cdist(_temp1, _temp2, p).squeeze().transpose(0, 1)
                _temp = torch.nn.functional.hardtanh(_temp, min_val=-1., max_val=1.)
                grad_X = torch.matmul(grad_output.transpose(0, 1), _temp)
            return grad_W, grad_X
    return cdist().apply
```

**Comment by ngimel** [comment]
This (or very similar) was implemented in #31167

**Comment by promach** [comment]
@ngimel are you saying that I should install the nightly version of pytorch for https://github.com/pytorch/pytorch/pull/31167 ?

@jacobrgardner As for https://github.com/pytorch/pytorch/pull/25799#issuecomment-529021810 , how to replace `torch.cat()` which contains [contiguous tensors](https://discuss.pytorch.org/t/understanding-cdist-function/76296/12?u=promach) ?

> x1_ = torch.cat([-2. * x1, x1_norm, x1_pad], dim=-1)
> x2_ = torch.cat([x2, x2_pad, x2_norm], dim=-1)

**Comment by ngimel** [comment]
You can install either nightly or 1.5.
